arch/riscv/kernel/irq.c

```diff
@@ -58,7 +58,7 @@ struct irq_chip riscv_irq_chip = {
 void __init init_IRQ(void)
 {
 	unsigned int irq;
-	for (irq = 0; irq < NR_IRQS; irq++)
+	for (irq = 0; irq < 2; irq++)
```

Comment:
> Another option that occurred to me is to specify the core IRQs in the device tree, when CONFIG_OF is defined.

That is certainly a possibility, but the present interpretation of `NR_IRQS_CORE` is inherently tied to the privileged ISA specification.  Any change would likely require extending `riscv_irq_{,un}mask()` as well, so in practice, it's probably not configurable as a device tree property would suggest.

> Moving the actual driver to the hypervisor seems to be at odds with
protecting the system from the driver. Anyway, I'm very interested in
thoughts on a new Driver Execution Environment.

One can imagine the drivers running in a separate S-mode environment, with H-mode acting as an intermediary.  Alternatively, platforms without H-mode could support a simple LPAR-like scheme in M-mode using base/bounds registers to denote accessible physical address ranges.